synapse/storage/databases/main/stream.py

```diff
@@ -770,7 +781,7 @@ def get_stream_id_for_event_txn(
         self,
         txn: LoggingTransaction,
         event_id: str,
-        allow_none=False,
+        allow_none: Literal[False] = False,
```

Comment:
This doesn't look right; get_stream_id_for_event_txn passes `allow_none=True`.  Could you explain your thinking?

Reply:
If is set `allow_none: bool = False,` I am get:
```
synapse/storage/databases/main/stream.py:786: error: No overload variant of "simple_select_one_onecol_txn" of "DatabasePool" matches argument types "LoggingTransaction", "str", "Dict[str, str]", "str", "bool"  [call-overload]
synapse/storage/databases/main/stream.py:786: note: Possible overload variants:
synapse/storage/databases/main/stream.py:786: note:     def simple_select_one_onecol_txn(cls, txn: LoggingTransaction, table: str, keyvalues: Dict[str, Any], retcol: str, allow_none: Literal[False] = ...) -> Any
synapse/storage/databases/main/stream.py:786: note:     def simple_select_one_onecol_txn(cls, txn: LoggingTransaction, table: str, keyvalues: Dict[str, Any], retcol: str, allow_none: Literal[True] = ...) -> Optional[Any]
```

I cannot see the mistake:
This should match with:
https://github.com/matrix-org/synapse/blob/98ec375b26ff167f91d40cf5e531d5640fca3635/synapse/storage/database.py#L1564-L1571

Reply:
My understanding:

1. `StreamWorkerStore.get_stream_id_for_event_txn` is declared to return an `int`.
2. `StreamWorkerStore.get_stream_id` returns whatever `DatabasePool.simple_select_one_onecol_txn` returns. The former passes a value of `allow_none` through to the latter.
3. `DatabasePool.simple_select_one_onecol_txn` returns `Optional[Any]`. If it's called with `allow_none=False`, then it returns `Any`; the `None` case is handled by raising `StoreError`. (See https://github.com/matrix-org/synapse/blob/e630722f11821b4d7610bceae256674114f9803b/synapse/storage/database.py#L1539-L1561; this is where the overloads mentioned in the error message come from).

Mypy has spotted that if we call `get_stream_id_for_event_txn` with `allow_none=True`, then `simple_select_one_onecol_txn` might return `None`. And that would contradict the return type annotation of `get_stream_id_for_event_txn`.

We only seem to call `get_stream_id_for_event_txn` in two places, both in this function:

https://github.com/matrix-org/synapse/blob/3e0536cd2afb5a640619bd872fc27b068ec3eb9b/synapse/storage/databases/main/event_push_actions.py#L184-L216

In particular line 194 of the quoted source looks like it's expecting `stream_ordering` to be possibly None.

Reply:
Given the above, I think the right thing to do is

1. change `get_stream_id_for_event_txn` to return `Optional[int]` instead of `int`.
2. Try changing `EventPushActionsWorkerStore` to inherit from `StreamWorkerStore` and removing the type-ignores where we call `get_stream_id_for_event_txn`. Doing so might fail due with DataStore MRO problems (related: #11165 and more generally #11733). If so, that's fine and we can leave things as they are.

Sorry Dirk, that's maybe a bit of information overload. Does that make sense? (If not, is there anything I can clarify?)

Reply:
Does it make sense to open an extra issue for this? And do not touch this type hint here?

Reply:
See #12437.

Reply:
Thank you for opening the new issue.